concurrency/DynamicFontDownloadHelper.java

```diff
@@ -0,0 +1,896 @@
+            for (String lang : fontInfo.downloadableFont.languageSet.getLanguages()) {
+                FontFamily fontFamily = Utils.getFactory().getUserSettingsController().getFontFamily(lang);
+                if (fontFamily.equals(fontInfo.getFontFamily())) {
+                    Utils.getFactory().getUserSettingsController().setFontFamily(FontFamily.getDefaultFont(lang), lang);
+                }
+            }
+
+            reportFontDownloadOnDemandMetrics(fontInfo, ReadingStreamUtil.ACTION_DOWNLOAD_ON_DEMAND_CLICK_DELETE_BUTTON);
+        }
+    }
+
+    /**
+     * An object to record the info of on demand downloaded fonts, including the downloading status & fonts.
+     */
+    public class OnDemandFontInfo implements IFontInfo {
+        private DownloadStatus downloadStatus;
+        private DownloadableFonts downloadableFont;
+
+        public OnDemandFontInfo(DownloadableFonts downloadableFont) {
+            this.downloadableFont = downloadableFont;
+            if (Utils.getFactory().getFontConfigInitializer().validateFont(this.downloadableFont.fontFamily)) {
+                this.downloadStatus = DownloadStatus.DOWNLOADED;
+            } else {
+                this.downloadStatus = DownloadStatus.NEED_DOWNLOAD;
+            }
+        }
+
+        @Override
+        public boolean isDownloaded() {
+            return downloadStatus == DownloadStatus.DOWNLOADED;
+        }
+
+        @Override
+        public boolean isDownloading() {
+            return downloadStatus == DownloadStatus.DOWNLOADING;
+        }
+
+        @Override
+        public String getResourceSizeString() {
+            return downloadableFont.resourceSize;
+        }
+
+        @Override
+        public int getDownloadStatusStringId() {
+            return 0;
+        }
+
+        public DownloadableFonts getDownloadableFont() {
+            return this.downloadableFont;
+        }
+
+        public void setDownloading() {
+            downloadStatus = DownloadStatus.DOWNLOADING;
+        }
+
+        public void setDownloaded() {
+            downloadStatus = DownloadStatus.DOWNLOADED;
+        }
+
+        @Override
+        public FontFamily getFontFamily() {
+            return this.downloadableFont.fontFamily;
+        }
+
+        @Override
+        public boolean isDefaultFont() {
+            return this.downloadableFont.defaultFont;
+        }
+    }
+
+    /**
+     * Get font info with font family.
+     * @param font
+     * @return
+     */
+    @CheckForNull
+    public OnDemandFontInfo getFontInfo(FontFamily font) {
+        if (font == null) {
+            return null;
+        }
+
+        OnDemandFontInfo fontInfo = fontInfoMap.get(font);
+        if (fontInfo == null) {
+            DownloadableFonts downloadableFonts = getDownloadableFonts(font);
+            if (downloadableFonts != null) {
+                fontInfo = new OnDemandFontInfo(downloadableFonts);
+                fontInfoMap.put(font, fontInfo);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **748, 749, and 753**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **756**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.